mtt/src/mtt_types.rs

```diff
@@ -118,3 +116,226 @@ impl MttL1EntryType {
+        }
+        _ => panic!("Attempt to get MttL1Dir from {l2_entry:?} for {addr_bits:x}"),
+    }
+}
+
+fn get_l1_index_and_subindex(phys_addr: SupervisorPageAddr) -> (u64, u64) {
+    let addr_bits = phys_addr.bits();
+    // Bits 25:17 for the 9-bit index, then 3-bit left shift for the 8-byte entry
+    let l1_index = (addr_bits & 0x03fe_0000) >> 14;
+    // 5-bit sub-index (16:12) with 1-bit left shift for the 2-bits per entry
+    let sub_index = (addr_bits & 0x1_f000) >> 11;
+    (l1_index, sub_index)
+}
+
+fn get_l1_value(pfn: u64, l1_index: u64) -> u64 {
+    let mtt_l1_page = pfn << 12;
+    let l1_entry_addr = (mtt_l1_page + l1_index) as *const u64;
+    assert!(l1_index <= 4096 - 8);
+    // Safety: The 4K L1 page is allocated by the TSM, and we are using a 12-bit index.
+    unsafe { *l1_entry_addr }
+}
+
+// Helper function to update value for an existing L1 entry
+fn update_l1_value(pfn: u64, phys_addr: SupervisorPageAddr, memory_type: MttMemoryType) {
+    let (l1_index, sub_index) = get_l1_index_and_subindex(phys_addr);
+    let value = get_l1_value(pfn, l1_index);
+    // Clear the original 2-bits and OR in the new value
+    let value = (value & !(3 << sub_index)) | ((memory_type as u64) << sub_index);
+    let mtt_l1_page = pfn << 12;
+    let l1_entry_addr = (mtt_l1_page + l1_index) as *mut u64;
+    assert!(l1_index <= 4096 - 8);
+    // Safety: The 4K L1 page is allocated by the TSM, and we are using a 12-bit index.
+    unsafe {
+        *l1_entry_addr = value;
+    };
+}
+
+// The L1 is a 4KB page that's allocated by the TSM to manage the memory type (non-confidential or confidential) at a 4K-page granularity.
+// This is done by dividing the 64MB region containing the page into 16KB regions of 4K each, and the 2-bit encoding in the L1 page pointed
+// to by the L2 entry determines the type.
+// The 12-bit index into the 4K page is computed using the 14-bits [25:12] of the physical address, and then shifting right by 2
+fn update_l1_entry(phys_addr: SupervisorPageAddr, memory_type: MttMemoryType) {
+    let addr = phys_addr.bits();
+    let l2_entry = get_l2_entry(phys_addr);
+    match l2_entry {
+        MttL1Dir(pfn) => update_l1_value(pfn, phys_addr, memory_type),
+        _ => panic!("Attempt to set MttL1Dir on {l2_entry:?} for {addr:x}"),
+    }
+}
+
+/// Changes the L2 type for 16 consecutive entries (each representing a 64MB range) for the 1GB range
+/// containing the address to the specified type.
+/// This can be called to consolidate previously fractured ranges.
+pub fn set_1gb_range_type(phys_addr: SupervisorPageAddr, memory_type: MttMemoryType) {
+    let gb_range_start = PageAddr::new(SupervisorPhysAddr::supervisor(
+        PageSize::Size1G.round_down(phys_addr.bits()),
+    ))
+    .unwrap();
+    for i in 0u64..=15 {
+        // Unwrap OK: We are starting with an aligned address and are adding 0 or 16K 4K pages (64MB)
+        let range_addr = gb_range_start.checked_add_pages(i * 16 * 1024).unwrap();
```

Comment:
Is the intent here: `for range_addr in gb_range_start.iter_from().take(16) {`?

Reply:
Not quite...the `for range_addr in gb_range_start.iter_from().take(16) {` will create an iterator of 16-elements of PageSize if I am not mistaken. In this case, the 1GB range is composed of 16 consecutive entries of 64MB each (of the same type by convention). So, assuming we starting with a 1GB range that starts with say 0, we'll increment by 16 4K pages (64MB, 128MB, and so forth). 

Perhaps, there's a better way of expressing it for clarity?

Reply:
will `step_by` work? https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.step_by

so `...iter_from().step_by(16).tak(16)`?

Reply:
Unfortunately, this results in in sub-optimal codegen in both release and debug per offline conversation. The most likely explanation is that we the step_by() actually needs to perform 16K passes per iteration (as opposed to a single multiply in the imperative version). This can likely be improved by iterator enhancements, but punting for now. 

The unwrap() calls have been replaced with an .ok_or().